Replace Webpack with Vite for client build

dannon opened the pull request:
Following up on the Vitest migration, this swaps out Webpack for Vite as the
client bundler. Since Vitest already uses Vite under the hood, this gets us on
a single modern toolchain for both building and testing.

The branch is a cleaned up rebase of previous Vite work - same functionality,
just reorganized into cleaner commits. Webpack and Babel dependencies are
removed (~30 packages), build time is around 35 seconds, and all the output
files match what templates expect.

Builds on #21194 and #21348

Will probably need some followup, but the dev server setup ended up being
straightforward - no crazy workarounds needed, and it should be easy to build
on going forward.


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

dannon committed:
Convert localization from AMD to ES modules

dannon committed:
Add Vite build scripts and dependencies to package.json

dannon committed:
Add Vite configuration and shared build config

dannon committed:
Add vite-plugin-webpack-aliases for legacy module resolution

dannon committed:
Add vite-plugin-build-metadata for Galaxy build info

dannon committed:
Add ESM-compatible jQuery wrapper for Vite

dannon committed:
Fix SCSS imports for Vite compatibility

Use explicit paths for SCSS imports that Vite cannot resolve, including custom_theme_variables.scss in App.vue, base.scss in 3 files, _breakpoints.scss in 7 files, and theme/blue.scss in 1 file.

dannon committed:
Fix web worker path for Vite compatibility

Vite requires relative paths for worker imports to resolve correctly.

dannon committed:
Fix CSS comment syntax for Vite compatibility

Replace // comments with /* */ which is valid CSS syntax.

dannon committed:
Fix FormGeneric icon prop type for Vite compatibility

Use Object type instead of IconDefinition for the icon prop to avoid
type import issues in non-TypeScript Vue files.

dannon committed:
Fix faGear import to use Font Awesome 6

faGear is only available in Font Awesome 6 (FA5 uses faCog), so import
it from the font-awesome-6 alias.

dannon committed:
Add Vite dev server with proxy support

Configure Vite to proxy all requests (except its own routes) to the Galaxy backend, matching webpack-dev-server behavior. Supports GALAXY_URL env var to specify backend (default: http://127.0.0.1:8080), CHANGE_ORIGIN env var for remote server support, and listens on 0.0.0.0 for network access.

dannon committed:
Formatting for locale/imports

dannon committed:
Add libs entry point for window globals

Creates src/entry/libs.js that exposes required globals (jQuery, underscore, bundleEntries, config) to window for legacy Mako templates. Also updates vite.config.mjs to add libs as an entry point and disables CSS code splitting for single base.css output.

dannon committed:
Switch build scripts to use Vite instead of webpack

dannon committed:
Remove webpack configuration and dependencies

dannon committed:
Clean up remaining webpack references

Rename vite-plugin-webpack-aliases.js to vite-plugin-galaxy-legacy.js and remove webpack-related comments and unused getWebpackAliases function.

dannon committed:
Remove remaining webpack references and dead code

Remove publicPath.js (webpack-specific), consolidate jquery.custom.js to ESM version, remove __webpack_public_path__ global, and update comments throughout to remove webpack mentions.

dannon committed:
Add @rollup/plugin-inject for jQuery/underscore globals

Replicates webpack's ProvidePlugin behavior by auto-injecting imports
for $, jQuery, and _ wherever they're used as bare identifiers. This
is needed because Vite outputs ES modules where imports are hoisted,
so window.jQuery assignments in libs.bundled.js execute after chunks
that depend on jQuery globals.

dannon committed:
Use ES module script tags for Vite output

Vite outputs ES modules, so script tags need type="module" to:
1. Load bundled JS files as modules (dist_js helper)
2. Ensure inline scripts execute after modules are loaded

Without type="module", inline scripts using globals like config or $
would execute before the module exports are available.

dannon committed:
Refactor entry points for Vite compatibility

jQuery no longer uses an alias in build-config.shared.js because Vite's code
splitting caused TDZ errors when jquery was aliased. It now resolves directly
from node_modules.

The jquery.custom.js wrapper is removed since it interfered with Rollup's
chunking. Instead, libs.js sets up window.$ and window.jQuery directly after
importing jquery, then uses a fire-and-forget dynamic import for jquery-migrate
to ensure proper load order.

bundleEntries and config exports are moved from libs.js to the app entry points
(analysis/index.ts and generic.js) to avoid circular dependency issues during
Vite code splitting.

dannon committed:
Add Node.js polyfills for browser compatibility

The events package is added because some dependencies (like Terminal which
extends EventEmitter) expect the Node.js events module to be available. Vite
doesn't automatically polyfill Node.js built-ins like webpack did, so we need
to provide this explicitly.

dannon committed:
Update Vite config for production builds

A d3v3CompatPlugin is added to fix D3 v3's ES module incompatibility. D3 v3
is an IIFE that expects 'this' to be 'window' at module level, but in ES
modules 'this' is undefined. The plugin transforms the IIFE to explicitly
call with the window context.

Buffer is injected via @rollup/plugin-inject since some dependencies expect
the Node.js Buffer to be globally available.

jQuery is split into its own chunk via manualChunks to ensure it loads before
jquery-migrate regardless of how Rollup decides to split the code.

The base is set to relative paths (./) so CSS asset references like fonts work
correctly when Galaxy is served behind a proxy prefix. The HTML script tags
use url_for() which handles the prefix, so they don't need this.

dannon committed:
Update documentation and code for Vite migration

The Makefile and client README now reference Vite instead of webpack, and the
dev server port is updated from 8081 to 5173 which is Vite's default.

Also removed the webpackChunkName magic comments from a few dynamic imports
since those were webpack-specific and Vite ignores them anyway.

dannon committed:
Add sourcemap support to Vite production builds

The GXY_BUILD_SOURCEMAPS environment variable is used by the build-production-maps
script but wasn't wired up to Vite's config. Now production builds will generate
sourcemaps when that env var is set.

dannon committed:
Fix SCSS deprecation warning noise in Vite build

The mixed-decls deprecation is now obsolete in newer Sass versions, so silencing
it just produces "obsolete" warnings. Removed that and added silencing for the
color-functions and global-builtin deprecations instead, since those are coming
from legacy code using lighten(), darken(), etc. that will need a separate
refactoring effort to update to the modern color.adjust() syntax.

dannon committed:
Remove Babel dependencies

Babel was for Jest, Vitest uses esbuild. None of this is needed anymore.

dannon committed:
Update caniuse-lite database

dannon committed:
Fix resizable.png path in WorkflowMinimap

dannon committed:
Keep app/* in same chunk to fix circular dependency warnings

dannon committed:
Remove requirejs dependency

It was only used for legacy AMD module loading which is no longer needed with Vite.

dannon committed:
Add Vite dev server plugin for HMR support

This plugin intercepts HTML responses from Galaxy and rewrites them so that the
browser loads source files from Vite instead of production bundles. It injects
the Vite HMR client and a config stub that queues config.set() calls until the
real module loads. This makes yarn develop work with any Galaxy backend.

dannon committed:
Wire up dev server plugin and fix dev mode compatibility

Adds the galaxyDevServerPlugin to enable HMR when running yarn develop. Also
adds process.env polyfill for Node modules that expect it, and an esbuild
plugin to fix D3 v3's IIFE this binding during dependency pre-bundling.

dannon committed:
Process queued config.set() calls in analysis entry point

When running in dev mode, inline scripts in Galaxy's HTML may call config.set()
before this module loads. The dev server plugin injects a stub that queues these
calls, and this change processes that queue so the configuration is applied.

dannon committed:
Support gzip-compressed responses and non-module script tags in dev server

Remote Galaxy servers like usegalaxy.org typically send gzip-compressed HTML
responses, and their templates use regular script tags rather than module
scripts. This updates the dev server plugin to:

- Decompress gzip responses before transforming HTML
- Match script tags with or without type="module" attribute

Tested working against usegalaxy.org with full HMR support.

dannon committed:
Remove now-unused polyfills.js file

dannon committed:
Remove unused bbi/ BigWig reader library

dannon committed:
Remove unused farbtastic color picker plugin

dannon committed:
Remove unused jquery.sparklines plugin

dannon committed:
Remove core-js and regenerator-runtime polyfills

dannon committed:
Remove stageLibs gulp task

With Vite, libraries import directly from node_modules rather than
from staged copies. The gulp stageLibs task is no longer needed.

dannon committed:
Split vendor chunks for better caching

Separate bootstrap-vue and fontawesome into their own chunks so they
can be cached independently from Galaxy code changes.

dannon committed:
Fix staging to include subdirectories

dannon committed:
Adjust vendor chunk splitting for bootstrap-vue compatibility

Bootstrap-vue checks Vue.version at module init time, which fails when
loaded as a separate chunk because ES modules can execute in parallel.
Keep these libraries in the main bundle for now - can revisit with
dynamic imports in a followup.

dannon committed:
Minor final vitest upgrades

dannon commented:
Looks like I need to address memory usage issues to make the build work in CI, similar to webpack.  I'll iterate on it.

dannon committed:
Bump CI node heap size for vite build

dannon committed:
Add space after hid in history items

dannon committed:
Fix Monaco editor integration for Vite

- Configure Monaco workers with Vite's ?worker syntax in CustomToolEditor
- Use lazy monacoInstance pattern in YamlJs.ts to avoid AMD loader issues
- Import CustomToolEditor statically (async loading incompatible with Vue Router 3.x + Vite)

The static import increases bundle size by ~7MB but is necessary until Vue Router 4 migration.

dannon committed:
Fix vue-tsc errors for Monaco worker imports

Add type declarations for Vite's ?worker suffix imports and use the
proper IDisposable type from monaco-types.

dannon committed:
Fix workflow editor auto-layout breaking in Vite production builds

Convert the dynamic import of the layout module to a static import.
The dynamic import was causing the app to reinitialize after clicking
auto-layout in production builds, clearing the workflow stores.

This increases the initial bundle size slightly (elkjs is now bundled
with the main chunk instead of being lazy-loaded), but ensures the
workflow editor functions correctly.

TODO: Investigate why dynamic imports in Vite production builds cause
this behavior, and consider alternative code-splitting approaches
like defineAsyncComponent or manual chunk configuration.

dannon committed:
Add explicit space between tool name and description in toolbox

Vite's template compiler doesn't preserve whitespace between inline
elements the way Webpack did, so the tool name and description were
running together. Adding an explicit space fixes this.

dannon committed:
Fix custom tools selenium test using stable Monaco editor selector

The .view-line elements inside Monaco editor get re-rendered as the
editor updates, causing StaleElementReferenceException. Use the stable
.monaco-editor container instead.

dannon committed:
Remove unused config.set queue processing code

The queue processing for config.set() calls was added to handle a race
condition in Vite dev mode, but no Mako templates actually use config.set().
This removes the dead code and simplifies the Window type declaration.

mvdbeek commented on lib/galaxy_test/selenium/test_custom_tools.py:
```diff
@@ -98,12 +98,13 @@ def paste_tool(self):
-        editor = self.wait_for_selector_visible(".monaco-editor div.view-line")
+        # Try finding Monaco editor and replace skeleton content
+        self.sleep_for(self.wait_types.UX_RENDER)  # Allow editor to initialize
+        # Use the stable .monaco-editor container, not .view-line which gets re-rendered
```
excellent fix!

mvdbeek committed:
Use `whitespace: "preserve"` compiler option

This seems cleaner ?

mvdbeek reviewed (approved):
This is awesome, and I think the build also nicely stays below the 4GB limit during the build phase, while completing in reasonable time (83 seconds for me).

mvdbeek committed:
Revert "Use `whitespace: "preserve"` compiler option"

This reverts commit 57a3aa6bcf4af88546403b09f2836bdd2eee26e7.

mvdbeek commented:
ahhhh, the makefile specifies `--max-old-space-size=3072` in NODE_OPTIONS, which is less than what we do in CI!

mvdbeek committed:
Bump --max-old-space-size to 4G in Makefile

dannon committed:
Merge remote-tracking branch 'upstream/dev' into vite-reorganized

dannon committed:
Remove bundleEntries and window.config exports

bundleEntries.js was deleted upstream and no templates use
window.bundleEntries or window.config anymore. Removes dead code
from entry points.

dannon committed:
Use whitespace: preserve to match Webpack vue-loader behavior

This preserves whitespace in Vue templates globally, matching the
default behavior of Webpack's vue-loader. Removes the need for
explicit {{ " " }} hacks in Tool.vue.

Note: Vue 3+ defaults to 'condense' which produces smaller bundles
but can break layouts when whitespace between elements spans
newlines. A future cleanup could switch to condense after auditing
components and using CSS margins instead of whitespace for spacing.
